Architecture overview (#54)
merantix-momentum/squirrel-core: docs/usage/architecture_overview.rst

* :ref:`catalog`: organizing, accessing, and sharing datasets.
* :ref:`driver`: performant and convenient read.
* :ref:`iterstream`: :py:class:`Composable` is the foundational building block in the :ref:`iterstream` module which provides a mechanism to chain iterables, and a fluent api that includes methods such as `map`, `filter`, and `async_map`.
* :ref:`store`: a key/value abstraction for reading data from and writing data to arbitrary storage backends such as filesystem, object store, database, etc.

These modules are designed in a way that `can` be
used together, but this is not enforced in order to maximize flexibility. This may make it difficult to realize the
intended and recommended way of combining squirrel primitives. Although there are many such ways already provided (and many more
that can be implemented for specific use-cases), here we focus on one concrete example that captures the most common
and most widely applicable pattern through a code snippet and its equivalent UML diagram.
Here is a complete data loading pipeline:

.. code-block:: python

    from squirrel.catalog import Catalog

>     catalog = Catalog.from_plugins() # Catalog
>     train_data = (
>         catalog["imagenet"] # CatalogSource

Reviewer: ```suggestion
    catalog = Catalog.from_plugins()  # Catalog
    train_data = (
        catalog["imagenet"]  # CatalogSource
```

streaming way for minimal memory footprint.

.. note::
      To access available datasets using `Catalog.from_plugins()`, check out `squirrel-dataset-core repository <https://github.com/merantix-momentum/squirrel-datasets-core>`_.

The following diagram illustrates a (simplified and slightly idealized) view of the relationships between these
classes through one concrete implementation provided by squirrel. Note that here we assume
that the data is in messagepack format (see :ref:`store` for information about different types of store).

.. mermaid::

    classDiagram

        MutableMapping <|-- Catalog
        class Catalog {
            Dict _sources
        }
        Catalog *-- "0..*" CatalogSource
        %% CatalogSource : get_driver()
        class CatalogSource {
            string identifier
            int version
            List~int~ versions

            get_driver() Driver
        }

        class MessagepackDriver {
            string name
            SquirrelStore store

            get(key) Iterable~Dict~
            keys() List~string~
            get_iter() Composable
        }

>         %% realiazation

Reviewer: ```suggestion
        %% realization
```

Architecture overview
=====================

There are four modules in squirrel which are instrumental in understanding the architecture and the overall design:

* :ref:`catalog`: organizing, accessing, and sharing datasets.
* :ref:`driver`: performant and convenient read.
* :ref:`iterstream`: :py:class:`Composable` is the foundational building block in the :ref:`iterstream` module which provides a mechanism to chain iterables, and a fluent api that includes methods such as `map`, `filter`, and `async_map`.
* :ref:`store`: a key/value abstraction for reading and writing data to and arbitrary storage backend such as filesystem, object store, database, etc.

These modules are designed in a way that `can` be
used together, but it is not enforced in order to maximize flexibility. This may make it difficult to realize the
intended and recommended way of combining squirrel primitives. Although there are many such ways already provided (and many more
that can be implemented for specific use-cases), here we focus on one concrete example that captures the most common
and most widely applicable pattern through a code snippet and its equivalent UML diagram.

> here is a complete data loading pipeline:

Reviewer 2: -> Here is a complete data loading pipeline:

.. code-block:: python

    from torch.utils.data import DataLoader
    from squirrel.catalog import Catalog

    catalog = Catalog.from_plugins() # Catalog

>     train_data = (
>         catalog["imagenet"] # CatalogSource
>         .get_driver()  # Driver
>         .get_iter()  # Composable
>         .map(lambda x: transform(x)) # Composable
>         .filter(lambda x: filter_func(x)) # Composable
>     )

Reviewer: ```suggestion
    train_data = (
        catalog["imagenet"]  # CatalogSource
        .get_driver()  # Driver
        .get_iter()  # Composable
        .map(lambda x: transform(x))  # Composable
        .filter(lambda x: filter_func(x))  # Composable
    )  # Composable
```

>     train_loader = DataLoader(train_data, batch_size=None)

Reviewer: Additional explanation and/or pointers for some lines here would be useful (some points I can think of are: what does from_plugins do, what is a plugin, where is "imagenet" coming from, what kind of funcs are `transform` and `filter_func`). Maybe converting the following paragraph to a step-by-step explanation format would be better?

Reviewer: I see, yes it is clear :+1:

A :py:class:`Catalog` contains zero to many :py:class:`CatalogSource`, each of which may be retrieved by a an
identifier (or a tuple of and identifier and the version, see :ref:`catalog` for more details). :py:class:`CatalogSource`

> contains all necessary information to instantiate an abject of type :py:class:`Driver`. :py:class:`Driver` may have

Reviewer 2: abject -> object

a method :py:meth:`get_iter` which returns an object of type :py:class:`Composable`
(which belongs to :ref:`iterstream` module). `train_data` is an iterable that generates items lazily in a
streaming way for minimal memory footprint.

.. note::
      To access available datasets using `Catalog.from_plugins()`, checkout `squirrel-dataset-core repository <https://github.com/merantix-momentum/squirrel-datasets-core>`_.

> The following diagram illustrating (a simplified and slightly idealized) view on the relationships between these

Reviewer 2: illustrating -> illustrates

> The following diagram illustrating (a simplified and slightly idealized) view on the relationships between these

Reviewer: ```suggestion
The following diagram illustrates a (simplified and slightly idealized) view of the relationships between these
```

        class MessagepackDriver {
            string name
            SquirrelStore store

            get(key) Iterable~Dict~
            keys() List~string~
            get_iter() Composable
        }

        %% realiazation
        CatalogSource ..|> MessagepackDriver

        MessagepackDriver ..> Composable
        MessagepackDriver ..> SquirrelStore

        <<abstract>> Composable
        class Composable {
            source Iterable~Any~
        }
        Composable : __iter__() Iterable~Any~
        Composable : map() Composable
        Composable : filter() Composable

        SquirrelStore : set(value, key) None
        SquirrelStore : get(key) Iterable~Any~
        SquirrelStore : keys() Iterable~string~

        SquirrelStore "1" --> MessagepackSerializer
        class MessagepackSerializer {
            serialize(obj)
            deserialize(obj)
            serialize_shard_to_file(obj, fp)
            deserialize_shard_from_file(fp)
        }

The relationships between these components and the methods they provide depends on the particular implementation of
the abstract classes (i.e. :py:class:`Drive`, :py:class:`AbstractStore`, :py:class:`SquirrelSerializer`).

> For instance, an implementation of the :py:class:`Drive` may not need or choose to not use :py:class:`SquirrelStore`

Reviewer 2: Spelling error in Driver: `the abstract classes (i.e. :py:class:`Driver`, :py:class:`AbstractStore`, :py:class:`SquirrelSerializer`).`

> For instance, an implementation of the :py:class:`Drive` may not need or choose to not use :py:class:`SquirrelStore`

Reviewer: ```suggestion
For instance, an implementation of the :py:class:`Drive` may not need to or may choose not to use :py:class:`SquirrelStore`
```
